subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublicationsTest.groovy

```diff
@@ -201,4 +207,37 @@ class DefaultConfigurationPublicationsTest extends Specification {
         explicit.artifacts == variantDef.artifacts
         explicit.children.empty
     }
+
+    def "can declare outgoing artifacts using lazy provider for configuration"() {
+        given:
+        artifactNotationParser.parseNotation(_ as DefaultFileSystemLocation) >> { args ->
+            def mockArtifact = Mock(ConfigurablePublishArtifact)
+            mockArtifact.getName() >> ((DefaultFileSystemLocation) args[0]).getAsFile().getName()
+            return mockArtifact
+        }
+        publications.setObjectFactory(TestUtil.objectFactory())
```

Comment:
Maybe a note here too that we indent to remove this mutator in favor of injection.